Start Guide for RAPIDS on AWS EMR 6.2 (#1371)
NVIDIA/spark-rapids: docs/get-started/getting-started-aws-emr.md

> # Get Started with Rapids on AWS EMR

Reviewer: ```suggestion
# Get Started with RAPIDS on AWS EMR

```

> This is a getting started guide for Rapids on AWS EMR. At the end of this guide, the user will be able to run a sample Apache Spark application that runs on NVIDIA GPUs on AWS EMR.

Reviewer: ```suggestion
This is a getting started guide for the RAPIDS Accelerator for Apache Spark on AWS EMR. At the end of this guide, the user will be able to run a sample Apache Spark application that runs on NVIDIA GPUs on AWS EMR.

```

> The current EMR 6.2.0 release support Spark version 3.0.1 and Nvidia spark-rapids 0.2.0. For more details of supported applications, please see the [EMR release notes](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-release-6x.html).  

Reviewer: ```suggestion
The current EMR 6.2.0 release supports Spark version 3.0.1 and RAPIDS Accelerator version 0.2.0. For more details of supported applications, please see the [EMR release notes](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-release-6x.html).  

```

--configurations file:///my-configurations.json \
--bootstrap-actions Name='My Spark Rapids Bootstrap action',Path=s3://my-bucket/my-bootstrap-action.sh
```

Please fill with actual value for KeyName and file paths. You can further customize  SubnetId, EmrManagedSlaveSecurityGroup, EmrManagedMasterSecurityGroup, name and region etc. 

###  Launch EMR Cluster using AWS Console (GUI)

Go to the AWS Management Console and select the `EMR` service from the "Analytics" section. Choose the region you want to launch your cluster in, e.g. US West Oregon, using the dropdown menu in the top right corner. Click `Create cluster` and select `Go to advanced options`, which will bring up a detailed cluster configuration page.

#### Step 1:  Software, Configuration and Steps

Select **emr-6.2.0** or latest EMR version for the release, uncheck all the software options, and then check **Hadoop 3.2.1**, **Spark 3.0.1**, **Livy 0.7.0** and **JupyterEnterpriseGateway 2.1.0**.

In the "Edit software settings" field, copy and paste the configuration from the [EMR document](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark-rapids.html). You can also create a JSON file on you own S3 bucket.

> ![Step 1: Step 1:  Software, Configuration and Steps](pics/Rapids_EMR_GUI_1.PNG)

Reviewer: Elsewhere in the `docs/` directory we're using `img` as the directory name to hold images.  It would be nice to keep this consistent.  Also RAPIDS should be all caps in the filenames.  Interesting it's the _only_ part of the image filenames that is not capitalized. :smile:

Reviewer 2: @mgzhao Tom pointed out that the config.json didn't have the right settings. I suggest that we provide a sample config.json here. For example; in this cluster you chose g4dn.2xlarge, which has 8 vCPU cores. The right config would be use all the cores, we need to spark.executor.cores=8, spark.task.resource.gpu.amount=0.125.
Here is the sample config for this cluster:
   {
      "Classification":"spark-defaults",
      "Properties":{
     	"spark.plugins":"com.nvidia.spark.SQLPlugin", 
     	"spark.sql.sources.useV1SourceList":"",
     	"spark.executor.extraJavaOptions":"-Dai.rapids.cudf.prefer-pinned=true",
     	"spark.executor.resource.gpu.discoveryScript":"/usr/lib/spark/scripts/gpu/getGpusResources.sh",      	
     	"spark.submit.pyFiles":"/usr/lib/spark/jars/xgboost4j-spark_3.0-1.0.0-0.2.0.jar",
     	"spark.executor.extraLibraryPath":"/usr/local/cuda/targets/x86_64-linux/lib:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/compat/lib:/usr/local/cuda/lib:/usr/local/cuda/lib64:/usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native:/docker/usr/lib/hadoop/lib/native:/docker/usr/lib/hadoop-lzo/lib/native",
     	"spark.rapids.sql.concurrentGpuTasks":"4",
     	"spark.executor.resource.gpu.amount":"1",
     	"spark.executor.cores":"8",
     	"spark.task.cpus ":"1",
     	"spark.task.resource.gpu.amount":"0.125",
     	"spark.rapids.memory.pinnedPool.size":"2G",
     	"spark.executor.memoryOverhead":"2G",
     	"spark.locality.wait":"0s",
     	"spark.sql.shuffle.partitions":"200",
     	"spark.sql.files.maxPartitionBytes":"256m",
      "spark.sql.adaptive.enabled":"false"
  	}

Reviewer 3: I updated the FAQ.md and created subdirectories in demos and img for different CSPs.  I also updated the file names.  I believe Kong had already updated the config settings in the json.

> ![Step 1: Step 1:  Software, Configuration and Steps](pics/Rapids_EMR_GUI_1.PNG)

Reviewer 4: some of hte configs in EMR guide are not right or should be adjusted for cluster setup.

SSH to the EMR cluster's master node, get into sparks shell and run the sql join example to verify GPU operation.

```
spark-shell
```

Running following Scala code in Spark Shell

```
val data = 1 to 10000
val df1 = sc.parallelize(data).toDF()
val df2 = sc.parallelize(data).toDF()
val out = df1.as("df1").join(df2.as("df2"), $"df1.value" === $"df2.value")
out.count()
```

### Submit Spark jobs to a EMR Cluster Accelerated by GPUs

Similar to spark-submit for on-prem clusters, AWS EMR supports a Spark application job to be submitted. The mortgage examples we use are also available as a spark application.  You can also use **spark shell** to run the scala code or **pyspark** to run the python code on master node through CLI.
 

### Running GPU Accelerated Mortgage ETL and XGBoost Example using EMR Notebook

An EMR Notebook is a "serverless" Jupyter notebook. Unlike a traditional notebook, the contents of an EMR Notebook itself—the equations, visualizations, queries, models, code, and narrative text—are saved in Amazon S3 separately from the cluster that runs the code. This provides an EMR Notebook with durable storage, efficient access, and flexibility.

> You can use the following step-by-step guide to run the example mortgage dataset using Rapids on Amazon EMR GPU clusters. For more examples, please refer to [NVIDIA/spark-rapids](https://github.com/NVIDIA/spark-rapids/)

Reviewer: This link brings the user to the main page for the RAPIDS Accelerator source code, not to more examples.  @sameerz @krajendrannv do you have a more appropriate link you'd like to see here?

Reviewer 2: yes @jlowe agreed. We need to point to the right ETL notebook and XGBoost Notebook

--configurations file:///my-configurations.json \
--bootstrap-actions Name='My Spark Rapids Bootstrap action',Path=s3://my-bucket/my-bootstrap-action.sh
```

Please fill with actual value for KeyName and file paths. You can further customize  SubnetId, EmrManagedSlaveSecurityGroup, EmrManagedMasterSecurityGroup, name and region etc. 

###  Launch EMR Cluster using AWS Console (GUI)

Go to the AWS Management Console and select the `EMR` service from the "Analytics" section. Choose the region you want to launch your cluster in, e.g. US West Oregon, using the dropdown menu in the top right corner. Click `Create cluster` and select `Go to advanced options`, which will bring up a detailed cluster configuration page.

#### Step 1:  Software, Configuration and Steps

Select **emr-6.2.0** or latest EMR version for the release, uncheck all the software options, and then check **Hadoop 3.2.1**, **Spark 3.0.1**, **Livy 0.7.0** and **JupyterEnterpriseGateway 2.1.0**.

In the "Edit software settings" field, copy and paste the configuration from the [EMR document](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark-rapids.html). You can also create a JSON file on you own S3 bucket.

> For cluster with 2x g4dn.2xlarge GPU instance as core/task nodes, we recommend following default settings

Reviewer 3: Updated the text.

The following steps are based on the AWS EMR document ["Using the Nvidia Spark-RAPIDS Accelerator
for Spark"](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark-rapids.html)

###  Launch an EMR Cluster using AWS CLI

You can use the AWS CLI to launch a cluster with one Master node (m5.xlarge) and two 
g4dn.2xlarge nodes: 

```
aws emr create-cluster \
--release-label emr-6.2.0 \
--applications Name=Hadoop Name=Spark Name=Livy Name=JupyterEnterpriseGateway \
--service-role EMR_DefaultRole \
--ec2-attributes KeyName=my-key-pair,InstanceProfile=EMR_EC2_DefaultRole \
--instance-groups InstanceGroupType=MASTER,InstanceCount=1,InstanceType=m4.4xlarge \
                  InstanceGroupType=CORE,InstanceCount=1,InstanceType=g4dn.2xlarge \
                  InstanceGroupType=TASK,InstanceCount=1,InstanceType=g4dn.2xlarge \
--configurations file:///my-configurations.json \
--bootstrap-actions Name='My Spark Rapids Bootstrap action',Path=s3://my-bucket/my-bootstrap-action.sh
```

Please fill with actual value for `KeyName` and file paths. You can further customize SubnetId,
EmrManagedSlaveSecurityGroup, EmrManagedMasterSecurityGroup, name and region etc. 

> The `my-configurations.json` installs the spark-rapids plugin on your cluster, configurs YARN to use

Reviewer: ```suggestion
The `my-configurations.json` installs the spark-rapids plugin on your cluster, configures YARN to use

```

> GPUs, configures Spark to use RAPIDS, and configures the YARN capacity scheduler.  An example json

Reviewer: ```suggestion
GPUs, configures Spark to use RAPIDS, and configures the YARN capacity scheduler.  An example JSON

```

		"Properties":{
			"enableSparkRapids":"true"
		}
	},
	{
		"Classification":"yarn-site",
		"Properties":{
			"yarn.nodemanager.resource-plugins":"yarn.io/gpu",
			"yarn.resource-types":"yarn.io/gpu",
			"yarn.nodemanager.resource-plugins.gpu.allowed-gpu-devices":"auto",
			"yarn.nodemanager.resource-plugins.gpu.path-to-discovery-executables":"/usr/bin",
			"yarn.nodemanager.linux-container-executor.cgroups.mount":"true",
			"yarn.nodemanager.linux-container-executor.cgroups.mount-path":"/sys/fs/cgroup",
			"yarn.nodemanager.linux-container-executor.cgroups.hierarchy":"yarn",
			"yarn.nodemanager.container-executor.class":"org.apache.hadoop.yarn.server.nodemanager.LinuxContainerExecutor"
		}
	},
	{
		"Classification":"container-executor",
		"Properties":{
			
		},
		"Configurations":[
			{
				"Classification":"gpu",
				"Properties":{
					"module.enabled":"true"
				}
			},
			{
				"Classification":"cgroups",
				"Properties":{
					"root":"/sys/fs/cgroup",
					"yarn-hierarchy":"yarn"
				}
			}
		]
	},
	{
        "Classification":"spark-defaults",
        "Properties":{
        "spark.plugins":"com.nvidia.spark.SQLPlugin",
        "spark.sql.sources.useV1SourceList":"",

>         "spark.executor.extraJavaOptions":"-Dai.rapids.cudf.prefer-pinned=true",

Reviewer 4: not needed anymore

        "spark.executor.resource.gpu.discoveryScript":"/usr/lib/spark/scripts/gpu/getGpusResources.sh",
        "spark.submit.pyFiles":"/usr/lib/spark/jars/xgboost4j-spark_3.0-1.0.0-0.2.0.jar",
        "spark.executor.extraLibraryPath":"/usr/local/cuda/targets/x86_64-linux/lib:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/compat/lib:/usr/local/cuda/lib:/usr/local/cuda/lib64:/usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native:/docker/usr/lib/hadoop/lib/native:/docker/usr/lib/hadoop-lzo/lib/native",

>         "spark.rapids.sql.concurrentGpuTasks":"4",

Reviewer 4: 4 might be a bit aggressive, 2 may be a safer default

When the cluster is ready, a green-dot will appear next to the cluster name and the "Status" column will display **Waiting, cluster ready**.

In the cluster's "Summary" tab, find the "Master public DNS" field and click the `SSH` button. Follow the instructions to SSH to the new cluster's master node.

![Finish Cluster Configuration](../img/AWS-EMR/RAPIDS_EMR_GUI_5.png)

### Running an example joint operation using Spark Shell

SSH to the EMR cluster's master node, get into sparks shell and run the sql join example to verify GPU operation.

```bash
spark-shell
```

Running following Scala code in Spark Shell

```scala
val data = 1 to 10000
val df1 = sc.parallelize(data).toDF()
val df2 = sc.parallelize(data).toDF()
val out = df1.as("df1").join(df2.as("df2"), $"df1.value" === $"df2.value")
out.count()
out.explain()
```

### Submit Spark jobs to a EMR Cluster Accelerated by GPUs

Similar to spark-submit for on-prem clusters, AWS EMR supports a Spark application job to be submitted. The mortgage examples we use are also available as a spark application.  You can also use **spark shell** to run the scala code or **pyspark** to run the python code on master node through CLI.
 

> 

Reviewer 4: extra newlines
